internal/characters/faruzan/burst.go

```diff
@@ -0,0 +1,176 @@
+	burstBuffKey  = "faruzan-q-dmg-bonus"
+	burstShredKey = "faruzan-q-shred"
+)
+
+func init() {
+	burstFrames = frames.InitAbilSlice(71)
+	burstFrames[action.ActionAttack] = 61 // Q -> N1
+	burstFrames[action.ActionAim] = 61    // Q -> Aim
+	burstFrames[action.ActionSkill] = 61  // Q -> E
+	burstFrames[action.ActionDash] = 61   // Q -> D
+	burstFrames[action.ActionJump] = 63   // Q -> J
+	burstFrames[action.ActionSwap] = 60   // Q -> Swap
+}
+
+// Faruzan deploys a Dazzling Polyhedron that deals AoE Anemo DMG and releases
+// a Whirlwind Pulse. While the Dazzling Polyhedron persists, it will
+// continuously move along a triangular path. Once it reaches each corner of
+// that triangular path, it will unleash 1 more Whirlwind Pulse.
+//
+// Whirlwind Pulse
+// - When the Whirlwind Pulse hits opponents, it will apply Perfidious Wind's
+// Ruin to them, decreasing their Anemo RES.
+// - The Whirlwind Pulse will also apply Prayerful Wind's Gift to all nearby
+// characters when it is unleashed, granting them Anemo DMG Bonus.
+func (c *char) Burst(p map[string]int) action.ActionInfo {
+	ai := combat.AttackInfo{
+		ActorIndex: c.Index,
+		Abil:       "The Wind's Secret Ways (Q)",
+		AttackTag:  combat.AttackTagElementalBurst,
+		ICDTag:     combat.ICDTagNone,
+		ICDGroup:   combat.ICDGroupDefault,
+		StrikeType: combat.StrikeTypeDefault,
+		Element:    attributes.Anemo,
+		Durability: 25,
+		Mult:       burst[c.TalentLvlBurst()],
+	}
+
+	c.Core.QueueAttack(
+		ai,
+		combat.NewCircleHit(c.Core.Combat.Player(), 6.3),
+		burstHitmark,
+		burstHitmark,
+		applyBurstShredCb,
+	)
+
+	// C2: The duration of the Dazzling Polyhedron created by
+	// The Wind's Secret Ways increased by 6s.
+	duration := 745
+	if c.Base.Cons >= 2 {
+		duration += 360
+	}
+
+	c.burstSrc = c.Core.F
+	currSrc := c.burstSrc
+
+	x, y := c.Core.Combat.Player().Pos()
+	count := 0
+	for i := 137; i <= duration; i += 120 {
+		ox, oy := calcGadgetOffsets(count)
+		c.Core.Tasks.Add(func() {
+			if c.burstSrc != currSrc {
+				return
+			}
+			for id := range c.Core.Combat.EnemiesWithinRadius(x+ox, y+oy, 6) {
+				trg, ok := c.Core.Combat.Enemy(id).(*enemy.Enemy)
+				if !ok {
+					continue
+				}
+				applyBurstShred(trg)
+			}
+		}, 43+i)
+		count += 1
+	}
+
+	// In-game refreshes 0.1s. We give buff every 69f to reduce spam.
+	field := combat.NewCircleHit(c.Core.Combat.Player(), 40)
+	for i := 0; i <= duration; i += 69 {
+		c.Core.Tasks.Add(func() {
+			if c.burstSrc != currSrc {
+				return
+			}
+			if !combat.WillCollide(field, c.Core.Combat.Player(), 0) {
+				return
+			}
+			for _, char := range c.Core.Player.Chars() {
+				c.applyBurstBuff(char)
+			}
+		}, 43+i)
+	}
+
+	// Last refresh to account for 0.1s tick period
+	c.Core.Tasks.Add(func() {
+		if c.burstSrc != currSrc {
+			return
+		}
+		if !combat.WillCollide(field, c.Core.Combat.Player(), 0) {
+			return
+		}
+		for _, char := range c.Core.Player.Chars() {
+			c.applyBurstBuff(char)
+		}
+	}, 43+int(duration/6.0)*6)
```

Comment:
You can simplify this to `43+duration`

Reply:
It was done this way since the refresh time is one frame off. For example, with the truncation math above, these would be the refresh points:

C2: 1105 -> 1104
C0: 765 -> 764

But I can see that this is slightly redundant since it's a one frame difference. I will be addressing this in the next set of commits.